cpu/stm32_common/periph/usbdev.c

```diff
@@ -0,0 +1,1168 @@
+#define STM32_USB_OTG_EP_TYPE_ISO       (0x01 << USB_OTG_DOEPCTL_EPTYP_Pos)
+#define STM32_USB_OTG_EP_TYPE_BULK      (0x02 << USB_OTG_DOEPCTL_EPTYP_Pos)
+#define STM32_USB_OTG_EP_TYPE_INTERRUPT (0x03 << USB_OTG_DOEPCTL_EPTYP_Pos)
+
+/* List of instantiated USB peripherals */
+static stm32_usb_otg_fshs_t _usbdevs[USBDEV_NUMOF] = { 0 };
+
+static usbdev_ep_t _out[_TOTAL_NUM_ENDPOINTS];
+static usbdev_ep_t _in[_TOTAL_NUM_ENDPOINTS];
+
+/* Forward declaration for the usb device driver */
+const usbdev_driver_t driver;
+
+static void _flush_fifo(const stm32_usb_otg_fshs_config_t *conf,
+                        uint8_t fifo_num);
+
+/*************************************************************************
+* Conversion function from the base address to specific register blocks *
+*************************************************************************/
+static USB_OTG_GlobalTypeDef *_global_regs(
+    const stm32_usb_otg_fshs_config_t *conf)
+{
+    return (USB_OTG_GlobalTypeDef *)(conf->periph + USB_OTG_GLOBAL_BASE);
+}
+
+static USB_OTG_DeviceTypeDef *_device_regs(
+    const stm32_usb_otg_fshs_config_t *conf)
+{
+    return (USB_OTG_DeviceTypeDef *)(conf->periph + USB_OTG_DEVICE_BASE);
+}
+
+static USB_OTG_INEndpointTypeDef *_in_regs(
+    const stm32_usb_otg_fshs_config_t *conf,
+    size_t endpoint)
+{
+    return (USB_OTG_INEndpointTypeDef *)(conf->periph +
+                                         USB_OTG_IN_ENDPOINT_BASE +
+                                         USB_OTG_EP_REG_SIZE * endpoint);
+}
+
+static USB_OTG_OUTEndpointTypeDef *_out_regs(
+    const stm32_usb_otg_fshs_config_t *conf,
+    size_t endpoint)
+{
+    return (USB_OTG_OUTEndpointTypeDef *)(conf->periph +
+                                          USB_OTG_OUT_ENDPOINT_BASE +
+                                          USB_OTG_EP_REG_SIZE * endpoint);
+}
+
+static __I uint32_t *_rx_fifo(const stm32_usb_otg_fshs_config_t *conf)
+{
+    return (__I uint32_t *)(conf->periph + USB_OTG_FIFO_BASE);
+}
+
+static __O uint32_t *_tx_fifo(const stm32_usb_otg_fshs_config_t *conf,
+                              size_t num)
+{
+    return (__O uint32_t *)(conf->periph +
+                            USB_OTG_FIFO_BASE +
+                            USB_OTG_FIFO_SIZE * num);
+}
+
+static __IO uint32_t *_pcgcctl_reg(const stm32_usb_otg_fshs_config_t *conf)
+{
+    return (__IO uint32_t *)(conf->periph + USB_OTG_PCGCCTL_BASE);
+}
+/* end of conversion functions */
+
+/**
+ * @brief Determine the number of available endpoints for the peripheral based
+ *        on the type and the CID version
+ *
+ * @param config    configuration struct
+ */
+static size_t _max_endpoints(const stm32_usb_otg_fshs_config_t *config)
+{
+    return (config->type == STM32_USB_OTG_FS) ?
+           STM32_USB_OTG_FS_NUM_EP :
+           STM32_USB_OTG_HS_NUM_EP;
+}
+
+static bool _uses_dma(const stm32_usb_otg_fshs_config_t *config)
+{
+#if STM32_USB_OTG_HS_USE_DMA
```

Comment:
```suggestion
#if STM32_USB_OTG_HS_ENABLED && STM32_USB_OTG_HS_USE_DMA
```

This will save you a lot of `#ifdef`s later on.